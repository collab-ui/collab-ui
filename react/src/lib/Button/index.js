/** @component button */

import React from 'react';
import PropTypes from 'prop-types';
import { Loading } from '@momentum-ui/react';
import omit from 'lodash/omit';

class Button extends React.Component {
  static displayName = 'Button';

  componentDidMount() {
    /* eslint-disable no-console */
    const { ariaLabel, ariaLabelledBy, index } = this.props;
    const { focusIndex, focusOnLoad } = this.context;
    (!ariaLabel && !ariaLabelledBy)
      &&
      console.warn('[@momentum-ui/react] Button: Accessibility could be improved with ariaLabel');

    focusOnLoad
    && focusIndex === index
    && this.refs.button.focus();
    /* eslint-enable no-console */
  }

  componentWillUpdate() {
    this.prevContext = this.context;
  }

  componentDidUpdate () {
    const { focusIndex } = this.context;
    const { index } = this.props;

    typeof index === 'number'
    && index === focusIndex
    && focusIndex !== this.prevContext.focusIndex
    && this.refs.button.focus();
  }

  handleKeyDown = e => {
    const { onClick, index } = this.props;
    const { handleKeyDown, handleClick } = this.context;
    if (e.which === 32 || e.which === 13 ||
        e.charCode === 32 || e.charCode === 13) {
      handleClick && handleClick(e, index);
      onClick && onClick(e);
    } else {
      handleKeyDown && handleKeyDown(e, index);
    }
  };

  handleClick = e => {
    const { handleClick } = this.context;
    const { index, onClick } = this.props;

    onClick && onClick(e);
    handleClick && handleClick(e, index);
  };

  render() {
    const {
      active,
      ariaLabel,
      ariaLabelledBy,
      children,
      circle,
      className,
      color,
      containerLarge,
      disabled,
      expand,
      href,
      index,
      isButtonGroup,
      label,
      loading,
      large,
      removeStyle,
      size,
      style,
      tag,
      type,
      ...props
    } = this.props;

    const otherProps = omit({...props}, [
      'onClick'
    ]);

    const { focusIndex } = this.context;

    const isButtonGroupIcon = () => (
      isButtonGroup
        && children
        && React.Children.toArray(children).reduce((prev, child) =>
          prev
            ? prev
            : child.type && child.type.displayName === 'Icon'
        , false)
    );

    const getChildren = () => {
      return (
        [
          loading &&
          <div key='child-0'>
            <Loading />
          </div>,
          <span
            className='md-button__children'
            style={{ opacity: `${loading ? 0 : 1}` }}
            key='child-1'
          >
            {children}
          </span>
        ]
      );
    };

    const getColor = () => {
      if (removeStyle) return false;

      return color === 'none' ? 'color-none' : color;
    };

    // Method for deprecated large prop
    const getSize = () => {
      if (removeStyle) return false;
      /* eslint-disable no-console */
      const validButtonSize = checkButtonSize();

      if (!circle && !validButtonSize) {
        console.warn('[@momentum-ui/react] Button: selected size is not supported for non-circular button. Size will default to 36');

        return '36';
      } else if (large) {
        console.warn('[@momentum-ui/react] Button: large prop is deprecated and will be removed. Please use size prop.');

        return !circle
          ? '52'
          : '44';

      } else {
        return size === 'none' ? 'size-none' : size;
      }
      /* eslint-enable no-console */
    };

    const checkButtonSize = () => (
      ['none', '28', '36', '40', '52', 28, 36, 40, 52].includes(size)
    );

    const button = React.createElement(
      tag,
      {
        ref: 'button',
        className:
          'md-button' +
          `${(circle && ` md-button--circle`) || ''}` +
          `${(isButtonGroupIcon() && ` md-button--icon-group`) || ''}` +
          `${(getSize() && ` md-button--${getSize()}`) || ''}` +
          `${(expand && ` md-button--expand`) || ''}` +
          `${(color && ` md-button--${getColor()}`) || ''}` +
          `${(removeStyle && ' md-button--none') || ''}` +
          `${(active && !disabled && ` active`) || ''}` +
          `${(className && ` ${className}`) || ''}`,
        onClick: this.handleClick,
        onKeyDown: this.handleKeyDown,
        style: style,
        disabled: disabled || loading,
        alt: ariaLabel || label,
        href: (tag === 'a' && href) || undefined,
        type: tag !== 'a' && type || '',
        ...ariaLabel
          ? { 'aria-label': ariaLabel }
          : { 'aria-labelledby': ariaLabelledBy },
        tabIndex: (typeof index !== 'number'
          || index === focusIndex)  ? 0 : -1,
        ...tag && tag !== 'button' && {role: 'button'},
        ...otherProps,
      },
      getChildren()
    );

    return (
      label
        ?
        <div
          className={`md-button__container${containerLarge ? '' : '--small'}`}
        >
          {button}
          <div className='md-button__label'>
            {label}
          </div>
        </div>
        :
        button
    );
  }
}

Button.contextTypes = {
  focusIndex: PropTypes.number,
  focusOnLoad: PropTypes.bool,
  handleClick: PropTypes.func,
  handleKeyDown: PropTypes.func,
};

Button.propTypes = {
  /** @prop Sets active css styling | false */
  active: PropTypes.bool,
  /** @prop Text to display for blindness accessibility features | '' */
  ariaLabel: PropTypes.string,
  /** @prop ID to reference for blindness accessibility feature | '' */
  ariaLabelledBy: PropTypes.string,
  /** @prop Children Nodes to Render inside Button | null */
  children: PropTypes.node.isRequired,
  /** @prop Sets circle css styling | false */
  circle: PropTypes.bool,
  /** @prop Optional css class string | '' */
  className: PropTypes.string,
  /** @prop Sets optional Button color | '' */
  color: PropTypes.string,
  /** @prop Sets containerLarge css styling | false */
  containerLarge: PropTypes.bool,
  /** @prop Sets the attribute disabled to Button | false */
  disabled: PropTypes.bool,
  /** @prop Sets expand css styling to widen the Button | false */
  expand: PropTypes.bool,
  /** @prop Href prop changes element to anchor element | '' */
  href: PropTypes.string,
  /** @prop This index is used to control focus of Button within a ButtonGroup | null */
  index: PropTypes.number,
  /** @prop Determines whether class should be applied to ButtonGroups with Icons as descendants | false */
  isButtonGroup: PropTypes.bool,
  /** @prop Text to display inside the button | '' */
  label: PropTypes.string,
  /** @prop Depreciated large css styling, use size instead | false */
  large: PropTypes.bool,
  /** @prop Activates the loading animation and sets the button to disabled | false */
  loading: PropTypes.bool,
  /** @prop Handler to be called when the user taps the button | null */
  onClick: PropTypes.func,
  /** @prop Optional prop to remove Button's default style | false */
  removeStyle: PropTypes.bool,
  /** @prop Optional string or number size prop | 36 */
  size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** @prop Additional css styling applied to the button | {} */
  style: PropTypes.object,
  /** @prop Optional tag to define type of element | 'button' */
  tag: PropTypes.oneOf(['button', 'input', 'a']),
  /** @prop Optional html type | 'button' */
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
};

Button.defaultProps = {
  active: false,
  ariaLabel: '',
  ariaLabelledBy: '',
  children: null,
  circle: false,
  className: '',
  color: '',
  containerLarge: false,
  disabled: false,
  expand: false,
  href: '',
  index: null,
  isButtonGroup: false,
  label: '',
  large: false,
  loading: false,
  onClick: null,
  removeStyle: false,
  size: 36,
  style: {},
  tag: 'button',
  type: 'button',
};

export default Button;
