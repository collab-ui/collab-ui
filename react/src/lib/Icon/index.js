/** @component icon */

import React from 'react';
import PropTypes from 'prop-types';
import find from 'lodash/find';
import trimEnd from 'lodash/trimEnd';
import trimStart from 'lodash/trimStart';
import iconNames from '@collab-ui/icons/data/iconNames.json';
import colors from '@collab-ui/core/data/colors.json';
import { Button } from '@collab-ui/react';

class Icon extends React.PureComponent {
  render() {
    const {
      ariaLabel,
      buttonClassName,
      color,
      className,
      description,
      isAria, // TODO(pajeter): remove isAria code with next major release
      onClick,
      name,
      size,
      sizeOverride,
      style,
      title,
      type,
      ...otherProps
    } = this.props;

    const consoleHandler = (message, data) => {
      /* eslint-disable no-console */
      switch (message) {
        case 'isAria-warn': // TODO(pajeter): remove isAria code with next major release
          console.warn(
            `[@collab-ui/react] Icon: isAria prop is deprecated and will be removed. Title, description or ariaLabel props should be used to add accessibility.`
          );
          break;
        case 'color-warn':
          console.warn(
            `[@collab-ui/react] Icon: ${data} may not exist in the design system,` +
              ` please use a color name from https://momentum.design/styles/color/style`
          );
          break;
        case 'color-error':
          console.warn(
            `[@collab-ui/react] Icon: ${data} does not exist in the design system,` +
              ` please use a color name from https://momentum.design/styles/color/style`
          );
          break;
          case 'name-error':
          console.warn(
            `[@collab-ui/react] Icon: Icon ${data} does not exist in the design system.` +
              ` Visit https://momentum.design/styles/icons/library for a list of available icons or to request a new icon.`
          );
          break;
      }
      /* eslint-enable no-console */
    };

    const getSize = () => {
      const defaultSize = 16;
      const sizeFromName = Number(name.split('_')[1]);
      return size || sizeFromName || defaultSize;
    };

    const getColorSpec = colorObj =>
      colorObj.hex
        ? colorObj.hex
        : colorObj.opacity && isolateRoot(colorObj.variable) === 'white'
          ? `rgba(255, 255, 255, ${colorObj.opacity / 100})`
          : `rgba(0, 0, 0, ${colorObj.opacity / 100})`;

    const getHexFromJSON = colorName => {
      for (let c of colors) {
        const variation = find(c.variations, ['variable', colorName]);

        if (variation) return getColorSpec(variation);
      }

      return consoleHandler('color-error', colorName);
    };

    const isolateRoot = str => {
      const trimmed = trimStart(str, '$');
      const split = trimmed.split('-');
      const root = split[0];
      return root;
    };

    const formatColor = () => {
      return color.startsWith('$')
        ? color
        : color.endsWith('-base')
          ? trimEnd(color, '-base')
          : `$${color}`;
    };

    const getColor = () => {
      if (!color) return 'inherit';

      if (color.startsWith('#')) {
        consoleHandler('color-warn', color);
        return color;
      }

      return getHexFromJSON(formatColor());
    };

    const getNameClass = () => {
      let iconName = name.startsWith('icon-')
        ? name.replace(/^(icon-)/,'')
        : name;

      if(sizeOverride) {
        iconName = name.split('_')[0] + `_${getSize()}`;
      }

      return iconNames.includes(iconName)
        ? `icon-${iconName}`
        : consoleHandler('name-error', iconName);
    };

    const styles = {
      fontSize: getSize(),
      color: getColor(),
      ...style
    };

    const getAriaLabel = () => {
      if (!isAria) return deprecationWarning(); // TODO(pajeter): remove isAria code with next major release
      if (isAria // TODO(pajeter): remove isAria code with next major release
        && ariaLabel) {
          return ariaLabel;
        }
      if (!ariaLabel) {
        if (title && description) return (`${title} ${description}`);
        if (title) return title;
        if (description) return description;
      }
      return null;
    };

    const deprecationWarning = () => { // TODO(pajeter): remove isAria code with next major release
      consoleHandler('isAria-warn');
    };

    const getIcon = () => {
      return (
        <i
          className={
            `cui-icon icon` +
            ` ${getNameClass()}` +
            `${(className && ` ${className}`) || ''}`
          }
          aria-label={!onClick ? getAriaLabel() : null}
          style={styles}
          {...!onClick && {...otherProps}}
        />
        );
    };

    return (
      onClick
        ?
        <Button
          className={
            'cui-button--icon' +
            `${(type && ` cui-button--icon-${type}`) || ''}` +
            `${(buttonClassName && ` ${buttonClassName}`) || ''}`
          }
          ariaLabel={getAriaLabel()}
          onClick={onClick}
          {...otherProps}
        >
          {getIcon()}
        </Button>
        :
        getIcon()
    );
  }
}

Icon.propTypes = {
  /** @prop Text to display for blindness accessibility features | null */
  ariaLabel: PropTypes.string,
  /** @prop Optional Button class name string | '' */
  buttonClassName: PropTypes.string,
  /** @prop Optional color css styling | '' */
  color: PropTypes.string,
  /** @prop Optional class name string | '' */
  className: PropTypes.string,
  /** @prop Icon description text | '' */
  description: PropTypes.string,
  /** @prop Depreciated prop that supports accessibility features | true */
  isAria: PropTypes.bool, // TODO(pajeter): remove isAria code with next major release
  /** @prop Required Icon name */
  name: PropTypes.string.isRequired,
  /** @prop Handler invoked by click of the user | null */
  onClick: PropTypes.func,
  /** @prop Sets Icon size | null */
  size: PropTypes.number,
  // Internal prop to override iconName with size prop */
  sizeOverride: PropTypes.bool,
  /** @prop Additional style properties to be applied | null */
  style: PropTypes.object,
  /** @prop Sets Icon Title prop | '' */
  title: PropTypes.string,
  /** @prop Sets Icon Type | '' */
  type: PropTypes.oneOf(['', 'white'])
};

Icon.defaultProps = {
  ariaLabel: null,
  buttonClassName: '',
  color: '',
  className: '',
  description: '',
  isAria: true, // TODO(pajeter): remove isAria code with next major release
  onClick: null,
  size: null,
  sizeOverride: false,
  style: null,
  title: '',
  type: ''
};

Icon.displayName = 'Icon';

export default Icon;
