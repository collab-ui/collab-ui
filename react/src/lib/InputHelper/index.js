/** @component input */

import React from 'react';
import PropTypes from 'prop-types';

const InputHelper = ({ message, className }) => {
  return <p className={`cui-input__help-text ${className}`}>{message}</p>;
};

InputHelper.propTypes = {
  /** @prop Optional css class name | '' */
  className: PropTypes.string,
  /** @prop Input help message for parent Input | null */
  message: PropTypes.string.isRequired
};

InputHelper.defaultProps = {
  className: '',
  message: null,
};

InputHelper.displayName = 'InputHelper';

export default InputHelper;

/**
* @component input
* @section Input Helper
*
* @react

import {
  Checkbox,
  InputHelper
} from '@collab-ui/react';

export default function InputHelperDefault() {
  return (
    <div className='row'>
      <Checkbox
        value='us'
        label='us'
        htmlId='inputHelper1'
        onChange={()=>{}}
      >
        <InputHelper message={'I\'m here to help'} />
      </Checkbox>
    </div>
  );
}

**/
