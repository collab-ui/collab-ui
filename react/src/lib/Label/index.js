/** @component label */

import React from 'react';
import PropTypes from 'prop-types';

const Label = ({ className, htmlFor, label, theme, ...props }) => {
  return (
    <label
      className={
        `${className ? ` ${className}` : ''}`+
        `${theme ? ` md-label--${theme}` : ''}`
      }
      htmlFor={htmlFor}
      {...props}
    >
      <span>{label}</span>
    </label>
  );
};

Label.propTypes = {
  /** @prop HTML class name for associated Input | '' */
  className: PropTypes.string,
  /** @prop HTML ID for associated Input | null */
  htmlFor: PropTypes.string.isRequired,
  /** @prop Required Label text | null */
  label: PropTypes.string.isRequired,
  /** @prop Set Label's color theme | '' */
  theme: PropTypes.string,
};

Label.defaultProps = {
  className: '',
  htmlFor: null,
  label: null,
  theme: '',
};

Label.displayName = 'Label';

export default Label;
