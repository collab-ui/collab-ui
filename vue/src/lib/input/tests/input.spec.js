import { mount } from '@vue/test-utils';
import Input from '../index.vue';
import Button from '../../button/index.vue';
import Icon from '../../icon/index.vue';
import InputError from '../../input-error/index.vue';
import InputHelper from '../../input-helper/index.vue';
import Label from '../../label/index.vue';
import Vue from 'vue';

Vue.component(Button.name, Button);
Vue.component(Icon.name, Icon);
Vue.component(InputError.name, InputError);
Vue.component(InputHelper.name, InputHelper);
Vue.component(Label.name, Label);

describe('Input', () => {
  it('should match snapshot', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: '1',
        name: 'test',
        label: 'test',
        type: 'text',
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should change tag based on multiline prop', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: '1',
        name: 'test',
        label: 'test',
        multiline: true,
      }
    });

    expect(wrapper.findAll('textarea').length).toEqual(1);
  });

  it('should match number SnapShot', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: '1',
        name: 'test',
        label: 'test',
        type: 'number',
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should match password SnapShot', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: '1',
        name: 'test',
        label: 'test',
        type: 'password',
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should match aria SnapShot', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: '1',
        name: 'test',
        label: 'test',
        ariaDescribedBy: 'ariaDescribedBy',
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should match clear SnapShot', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: '1',
        name: 'test',
        label: 'test',
        text: 'text',
        clear: true
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should render one Input', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: '1',
        name: 'test',
        label: 'test',
      }
    });

    expect(wrapper.findAll('input').length).toEqual(1);
    expect(wrapper.findAll('label').length).toEqual(1);
  });

  it('should render one Input and apply dirty class', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: '1',
        name: 'test',
        label: 'test',
        value: 'test'
      }
    });

    expect(wrapper.contains('.dirty')).toBe(true)
  });

  it('should render Label', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: '1',
        name: 'test',
        label: 'test',
      }
    });

    expect(
      wrapper.find('label').html()
    ).toEqual('<label for="1" class="md-label"><span>test</span></label>');
  });

  it('should pass class based on inputSize prop', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        inputSize: 'medium-12'
      }
    });

    expect(wrapper.classes('medium-12')).toBeTruthy();
    expect(wrapper.classes('columns')).toBeTruthy();
  });

  it('should pass placeholder attribute', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        placeholder: 'test'
      }
    });

    expect(wrapper.find('input').attributes().placeholder).toEqual('test');
  });

  it('should pass disabled attribute', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        disabled: true
      }
    });

    expect(wrapper.find('input').attributes().disabled).toEqual('disabled');
  });

  it('should pass readonly attribute', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        readonly: true
      }
    });

    expect(wrapper.find('input').attributes().readonly).toEqual('readonly');
  });

  it('should pass value attribute', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        value: 'testing'
      }
    });

    expect(wrapper.find('input').attributes().value).toEqual('testing');
  });

  it('should pass class based on nesting', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        nestedLevel: 1
      }
    });

    expect(wrapper.classes('md-input--nested-1')).toBeTruthy();
  });

  it('should render Secondary Container and Label', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        secondaryLabel: 'test'
      }
    });

    expect(wrapper.contains('.md-label__secondary-label-container')).toBeTruthy();
    expect(wrapper.contains('.md-label__secondary-label')).toBeTruthy();
  });

  it('should render Helper Text', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        inputHelpText: 'test'
      }
    });

    expect(wrapper.contains(InputHelper)).toBeTruthy();
  });

  it('should determine correct error class(warning)', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        errorArr: [{ error: 'test', type: 'warning' }]
      }
    });

    expect(wrapper.find('.md-input-group').classes('warning')).toBeTruthy();
  });

  it('should determine correct error class(error)', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        errorArr: [{ error: 'test', type: 'error' }]
      }
    });

    expect(wrapper.find('.md-input-group').classes('error')).toBeTruthy();
  });

  it('should determine correct error class if passed error and warning(error)', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        errorArr: [
          { error: 'error1', type: 'error' },
          { error: 'error2', type: 'warning' },
        ]
      }
    });

    expect(wrapper.find('.md-input-group').classes('error')).toBeTruthy();
  });

  it('should render Error Component', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        errorArr: [{ error: 'test', type: 'error' }]
      }
    });

    expect(wrapper.contains(InputError)).toBeTruthy();
  });

  it('should support input ref', () => {
    const inputRef = 'input1';
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        inputRef: inputRef
      }
    });
    expect(wrapper.vm.$refs[inputRef].tagName).toEqual('INPUT');
  });

  it('should render children', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
      },
      slots: {
        default: '<div class="testingforhoc" />',
      }
    });

    expect(wrapper.findAll('.testingforhoc').length).toEqual(1);
  });

  it('should handle change event', () => {
    let count = 0;
    const countUp = () => count++;
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
      },
      listeners: {
        change: countUp
      }
    });

    wrapper.vm.$emit('change', { target: { value: 'test' } });
    expect(count).toEqual(1);
  });

  it('should handle mouse down event', () => {
    let count = 0;
    const countUp = () => count++;
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
      },
      listeners: {
        mousedown: countUp
      }
    });

    wrapper.vm.$emit('mousedown');
    expect(count).toEqual(1);
  });

  it('should handle focus event', () => {
    let count = 0;
    const countUp = () => count++;
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
      },
      listeners: {
        focus: countUp
      }
    });

    wrapper.vm.$emit('focus');
    expect(count).toEqual(1);
  });

  it('should not render clear icon', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
      }
    });

    wrapper.vm.$emit('change', { target: { value: 'test' } });
    expect(wrapper.find('.md-button--icon').exists()).toBeFalsy();
  });

  it('should render clear icon if prop is present', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        name: 'test',
        label: 'test',
        value: 'test',
        clear: true
      }
    });

    expect(wrapper.find('.md-button--icon').exists()).toBeTruthy();
  });

  it('should clear value if clear icon is clicked', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        label: 'test',
        value: 'test',
        clear: true
      }
    });

    expect(wrapper.find('input').attributes().value).toEqual('test');
    wrapper.find('button.md-button--icon').trigger('click');

    expect(wrapper.find('input').attributes().value).toEqual('');
    expect(wrapper.find('.md-button--icon').exists()).toBeFalsy();
  });

  it('should clear value if Enter is pressed on the keyboard', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        label: 'test',
        value: 'test',
        clear: true
      }
    });

    expect(wrapper.find('input').attributes().value).toEqual('test');
    wrapper.find('button.md-button--icon').trigger('keydown.enter');
    expect(wrapper.find('input').attributes().value).toEqual('');
    expect(wrapper.find('.md-button--icon').exists()).toBeFalsy();
  });

  it('should clear value if Space is pressed on the keyboard', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        label: 'test',
        value: 'test',
        clear: true
      }
    });

    expect(wrapper.find('input').attributes().value).toEqual('test');
    wrapper.find('button.md-button--icon').trigger('keydown.space');
    expect(wrapper.find('input').attributes().value).toEqual('');
    expect(wrapper.find('.md-button--icon').exists()).toBeFalsy();
  });

  it('should focus on input when clear is triggered with htmlId', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test123',
        label: 'test',
        value: 'test',
        clear: true
      }
    });

    wrapper.find('button.md-button--icon').trigger('click');
    expect(wrapper.find('input') === document.activeElement);
  });

  it('should focus on input when clear is triggered with id', () => {
    const wrapper = mount(Input, {
      propsData: {
        id: 'test123',
        label: 'test',
        value: 'test',
        clear: true
      }
    });

    wrapper.find('button.md-button--icon').trigger('click');
    expect(wrapper.find('input') === document.activeElement);
  });

  it('should not render custom icon', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test',
        label: 'test',
        value: 'test',
      }
    });

    expect(wrapper.find('.md-input__icon').exists()).toBeFalsy();
  });

  it('should render custom icon if prop is present', () => {
    const wrapper = mount(Input, {
      propsData: {
        htmlId: 'test',
        label: 'test',
        value: 'test',
        clearIconName: 'icon-info_16'
      }
    });

    expect(wrapper.find('.md-icon').exists()).toBeTruthy();
  });

});

