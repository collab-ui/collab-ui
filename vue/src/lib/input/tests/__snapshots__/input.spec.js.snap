// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match aria SnapShot 1`] = `<div class="md-input-group"><label for="1" class="md-label"><span>test</span></label><input tabindex="0" type="text" value="" id="1" aria-describedby="ariaDescribedBy" placeholder="" class="md-input"></div>`;

exports[`Input should match clear SnapShot 1`] = `
<div class="md-input-group" text="text"><label for="1" class="md-label"><span>test</span></label>
  <div class="md-input__icon-container"><input tabindex="0" type="text" value="" id="1" placeholder="" class="md-input"></div>
</div>
`;

exports[`Input should match number SnapShot 1`] = `<div class="md-input-group"><label for="1" class="md-label"><span>test</span></label><input tabindex="0" type="number" value="" id="1" placeholder="" class="md-input"></div>`;

exports[`Input should match password SnapShot 1`] = `<div class="md-input-group"><label for="1" class="md-label"><span>test</span></label><input tabindex="0" type="password" value="" id="1" placeholder="" class="md-input"></div>`;

exports[`Input should match snapshot 1`] = `<div class="md-input-group"><label for="1" class="md-label"><span>test</span></label><input tabindex="0" type="text" value="" id="1" placeholder="" class="md-input"></div>`;
